### PR TITLE
Improve the nbs decoder so it does not need the VM to run

### PR DESCRIPTION
**Pull request by TrentHouliston**
Improve the nbs decoder so that it can run without the aid of the VM. This opens up more opportunities to use it for manipulating nbs files with less effort.
Also add on code to synchronize on packet headers for malformed or partial nbs streams.

### CODE REVIEW

File: tools/nbs_decoder/__init__.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                        packet = (decoders[type_hash][0], timestamp, decoders[type_hash][1].FromString(payload[16:]))
+                        yield packet
+                    except:
+                        pass
+            else:
+                c = f.read(1)
+                # No bytes to read we are EOF
+                if len(c) == 0:
+                    sync_state = state.eof
+                elif c == b"\xE2":
```
Comment (Bidski) [inline tools/nbs_decoder/__init__.py]: State machine does not require a header to exist in its entirety

File: tools/nbs_decoder/__init__.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+proto_files = [
+    *glob.glob(os.path.join(nuclear_proto_dir, "**", "*.proto"), recursive=True),
+    *glob.glob(os.path.join(user_proto_dir, "**", "*.proto"), recursive=True),
+]
+
+# Build the protocol buffers
+if (
+    os.system(
+        "protoc --python_out={protobuf_path} -I{nuclear_dir} -I{user_dir} {proto_files}".format(
```
Comment (Bidski) [inline tools/nbs_decoder/__init__.py]: We should do a version check on protoc to ensure it meets a minimum version requirement (if we have one)

File: nuclear/message/Neutron.cmake [hunk 3]
```diff
@@ -164,10 +164,8 @@ FOREACH(proto ${protobufs})
     ADD_CUSTOM_COMMAND(
         OUTPUT "${outputpath}/${file_we}.pb.cc"
                "${outputpath}/${file_we}.pb.h"
-               "${outputpath}/${file_we}_pb2.py"
         COMMAND ${PROTOBUF_PROTOC_EXECUTABLE}
         ARGS --cpp_out=lite:${message_binary_include_dir}
-             --python_out=${message_binary_include_dir}
```
Comment (Bidski) [inline nuclear/message/Neutron.cmake]: Why is removed? This will be needed for python modules (eventually)

Comment (Bidski) [reply nuclear/message/Neutron.cmake]: Nevermind this one

File: tools/nbs_decoder/__init__.py (hunk 1 above)
Comment (TrentHouliston) [reply tools/nbs_decoder/__init__.py]: Your critique here is invalid as it was never a state machine in the first place.
On an unrelated note, I've made it a state machine now so it actually works :P

File: tools/nbs_decoder/__init__.py (hunk 2 above)
Comment (TrentHouliston) [reply tools/nbs_decoder/__init__.py]: That's what the if is for here. If protoc can't parse it then it'll return an error code which means we can't use it. The shell is not redirected so you'll see whatever protoc complained about in your terminal.

Comment (Bidski) [reply tools/nbs_decoder/__init__.py]: Its not a question of whether or not protoc will error on a "bad" file, but whether the error protoc will give will be ambiguous in the context

File: tools/nbs_decoder/__init__.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                        packet = (decoders[type_hash][0], timestamp, decoders[type_hash][1].FromString(payload[16:]))
+                        yield packet
+                    except:
+                        pass
+            else:
+                c = f.read(1)
+                # No bytes to read we are EOF
+                if len(c) == 0:
+                    sync_state = state.eof
+                elif c == b"\xE2":  # No need to check the state here as we can step to lock2 from any state
```
Comment (Bidski) [inline tools/nbs_decoder/__init__.py]: step to lock1?

File: tools/nbs_decoder/__init__.py (hunk 2 above)
Comment (TrentHouliston) [reply tools/nbs_decoder/__init__.py]: I feel like it's just an extra check that's not really needed for a simple tool

**Comment by TrentHouliston** [comment]
> Add an extra format file to the roles folder so we can specialise the formatting there

I think this was for #259

**Pull request merged**